### PR TITLE
fix LHERunInfoProduct duplication

### PR DESCRIPTION
**Pull request by perrozzi**
not having received any specific comment I tried to implement it in the most "general"way.
In my local tests it worked for aMG_aMC@NLO in LO and NLO mode, POWHEG with and without JHU and MiNLO.
Surely can be improved code-wise, let me know if you have comments (like what is the best way to pass to "find" an array of strings and get the OR of the match, more than duplicating the lines)

### CODE REVIEW

**Comment by perrozzi** [comment]
please test

**Comment by perrozzi** [comment]
+1

File: SimDataFormats/GeneratorProducts/src/LHERunInfoProduct.cc [hunk 1]
```diff
@@ -206,104 +206,102 @@ bool HeaderLess::operator() (const LHERunInfoProduct::Header &a,
...
 		// and test the rest against a whitelist
 
 		std::set<Header, HeaderLess> headers;
 		std::copy(headers_begin(), headers_end(),
 		          std::inserter(headers, headers.begin()));
 
+    // make a list of headers contained in the second file
+    std::vector<std::vector<std::string> > runcard_v2;
+    std::vector<std::string> runcard_v2_header;
+    for(std::vector<LHERunInfoProduct::Header>::const_iterator
```
Comment (kpedro88) [inline SimDataFormats/GeneratorProducts/src/LHERunInfoProduct.cc]: could be range-based for loop

File: SimDataFormats/GeneratorProducts/src/LHERunInfoProduct.cc [hunk 2]
```diff
@@ -206,104 +206,102 @@ bool HeaderLess::operator() (const LHERunInfoProduct::Header &a,
...
 				continue;
 			}
-
+            
+      if(header->tag()=="MGRunCard" || header->tag() == "mgruncard" || header->tag() == ""){
+        bool header_compatible = false;
+        for (unsigned int iter_runcard = 0; iter_runcard < runcard_v2.size(); iter_runcard++){
+          // this is relevant for MG5_aMG@NLO in LO mode
+          std::vector<std::string> runcard_v1 = header->lines();
+          runcard_v1.erase( std::remove_if( runcard_v1.begin(), runcard_v1.end(), [](const std::string& x) { return x.find("iseed") != std::string::npos; } ), runcard_v1.end() );
```
Comment (kpedro88) [inline SimDataFormats/GeneratorProducts/src/LHERunInfoProduct.cc]: StackOverflow suggests this list of `find`s could be replaced with one `find_if`:
http://stackoverflow.com/questions/19952155/pass-multiple-strings-to-the-stringfind-function/19952425#19952425

The `if` statements with multiple string equality comparisons could be replaced similarly.

**Comment by perrozzi** [comment]
@kpedro88 range-based for loop should be ok, concerning find_if I tried but couldn't figure out a proper and compact way of using it inside this nested line, so I come up with a simpler solution to compress the code. would it be acceptable?

@cmsbuild please test

**Comment by perrozzi** [comment]
please test

**Comment by perrozzi** [comment]
+1

**Comment by kpedro88** [comment]
@perrozzi your change in 694390f is more computationally efficient, but still results in unnecessary code duplication. You can follow exactly the [StackOverflow post](https://stackoverflow.com/questions/19952155/pass-multiple-strings-to-the-stringfind-function/19952425#19952425) I linked before:
```
std::vector<string> checklist{"iseed","Random",".log",".dat",".lhe"};
runcard_v1.erase( std::remove_if( runcard_v1.begin(), runcard_v1.end(), [&](const std::string& x) 
    { return checklist.end() != std::find_if(checklist.begin(),checklist.end(),[&](const std::string& y)
        { return x.find(y) != std::string::npos; } )
    } ),
    runcard_v1.end()
);
```

But I admit, the nested expression becomes a bit opaque. Perhaps a simpler solution would be to make the comparison lambda into a standalone function in an anonymous namespace, so it could be reused for both `remove_if` statements (instead of redefining the lambda each time). A similar approach could be used for another repeated comparison:
```
if(header2.tag()=="MGRunCard" || header2.tag() == "mgruncard" || header2.tag() == "")
```

I want to avoid the possibility of a future developer making inconsistent changes in the list of strings to compare.

**Comment by perrozzi** [comment]
I made an attempt and created a couple of vectors and a function called find_if_checklist, used to remove the lambda functions and replace the isTagComparedInMerge. It seems to work ok, please check.

However if the structure is OK there is another problem: if I want to remove isTagComparedInMerge 
then GeneratorInterface/LHEInterface/plugins/LHEProvenanceHelper.cc
complains because it uses this function.  I am not sure how to proceed to let LHEProvenanceHelper.cc know about the vectors in LHERunInfoProduct.cc (I should define them in the header, but then I could not initialize them).
Probably there is an easy solution but I am not getting it, please let me know if you have ideas.
Or I restore isTagComparedInMerge since is starting to become urgent.

**Comment by kpedro88** [comment]
@perrozzi you could just do:
```
bool LHERunInfoProduct::isTagComparedInMerge(const std::string& tag) {
        return !find_if_checklist(header->tag(),tag_checklist);
}
```

but I think you can also just leave `isTagComparedInMerge()` alone... I was asking you to replace these comparisons:
https://github.com/perrozzi/cmssw/blob/d38d92c54b7b27e02719a30fdc452f84ab7e96ef/SimDataFormats/GeneratorProducts/src/LHERunInfoProduct.cc#L249
https://github.com/perrozzi/cmssw/blob/d38d92c54b7b27e02719a30fdc452f84ab7e96ef/SimDataFormats/GeneratorProducts/src/LHERunInfoProduct.cc#L265

**Comment by perrozzi** [comment]
@kpedro88 I changed lines 249 and 265.
I had also tried the snippet you wrote in this form
https://github.com/cms-sw/cmssw/pull/18846/commits/b5b5cd849b918ce41528e68105ab6b4a04887b57#diff-bb2124e47b0d956b141c59f69f2496d4R221
but if I try to compile I get the error below [*] and this is where I had stopped,
let's see if you have a way out

thanks a lot
luca

[*]
```
>> Compiling  /afs/cern.ch/work/p/perrozzi/private/git/Hbb/validations/LHERunInfoProduct/CMSSW_9_2_0_patch1/src/SimDataFormats/GeneratorProducts/src/LHERunInfoProduct.cc
/afs/cern.ch/work/p/perrozzi/private/git/Hbb/validations/LHERunInfoProduct/CMSSW_9_2_0_patch1/src/SimDataFormats/GeneratorProducts/src/LHERunInfoProduct.cc: In static member function 'static bool LHERunInfoProduct::isTagComparedInMerge(const string&)':
/afs/cern.ch/work/p/perrozzi/private/git/Hbb/validations/LHERunInfoProduct/CMSSW_9_2_0_patch1/src/SimDataFormats/GeneratorProducts/src/LHERunInfoProduct.cc:222:52: error: cannot call member function 'bool LHERunInfoProduct::find_if_checklist(std::string, std::vector<std::basic_string<char> >)' without object
         return !find_if_checklist(tag,tag_checklist);
                                                    ^
config/SCRAM/GMake/Makefile.rules:2017: recipe for target 'tmp/slc6_amd64_gcc530/src/SimDataFormats/GeneratorProducts/src/SimDataFormatsGeneratorProducts/LHERunInfoProduct.o' failed
gmake: *** [tmp/slc6_amd64_gcc530/src/SimDataFormats/GeneratorProducts/src/SimDataFormatsGeneratorProducts/LHERunInfoProduct.o] Error 1
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2

```

**Comment by kpedro88** [comment]
Ah, I didn't realize `isTagComparedInMerge()` was a static function. My suggestion won't work, just leave it the way it was.

**Comment by perrozzi** [comment]
please test

**Comment by perrozzi** [comment]
it should be ok now

**Comment by perrozzi** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**